In [36]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("olistbr/brazilian-ecommerce")

print("Path to dataset files:", path)

Using Colab cache for faster access to the 'brazilian-ecommerce' dataset.
Path to dataset files: /kaggle/input/brazilian-ecommerce


In [37]:
import pandas as pd
import os

# The path variable from the previous cell holds the directory where the datasets are located.
# path = '/kaggle/input/brazilian-ecommerce'

customers = pd.read_csv(os.path.join(path, "olist_customers_dataset.csv"))
orders = pd.read_csv(os.path.join(path, "olist_orders_dataset.csv"))
items = pd.read_csv(os.path.join(path, "olist_order_items_dataset.csv"))
payments = pd.read_csv(os.path.join(path, "olist_order_payments_dataset.csv"))
products = pd.read_csv(os.path.join(path, "olist_products_dataset.csv"))

print("Datasets loaded successfully!")
print(f"Customers shape: {customers.shape}")
print(f"Orders shape: {orders.shape}")
print(f"Items shape: {items.shape}")
print(f"Payments shape: {payments.shape}")
print(f"Products shape: {products.shape}")

Datasets loaded successfully!
Customers shape: (99441, 5)
Orders shape: (99441, 8)
Items shape: (112650, 7)
Payments shape: (103886, 5)
Products shape: (32951, 9)


In [26]:
customers = customers
customers

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP
...,...,...,...,...,...
99436,17ddf5dd5d51696bb3d7c6291687be6f,1a29b476fee25c95fbafc67c5ac95cf8,3937,sao paulo,SP
99437,e7b71a9017aa05c9a7fd292d714858e8,d52a67c98be1cf6a5c84435bd38d095d,6764,taboao da serra,SP
99438,5e28dfe12db7fb50a4b2f691faecea5e,e9f50caf99f032f0bf3c55141f019d99,60115,fortaleza,CE
99439,56b18e2166679b8a959d72dd06da27f9,73c2643a0a458b49f58cea58833b192e,92120,canoas,RS


In [27]:
orders = orders
orders

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00
...,...,...,...,...,...,...,...,...
99436,9c5dedf39a927c1b2549525ed64a053c,39bd1228ee8140590ac3aca26f2dfe00,delivered,2017-03-09 09:54:05,2017-03-09 09:54:05,2017-03-10 11:18:03,2017-03-17 15:08:01,2017-03-28 00:00:00
99437,63943bddc261676b46f01ca7ac2f7bd8,1fca14ff2861355f6e5f14306ff977a7,delivered,2018-02-06 12:58:58,2018-02-06 13:10:37,2018-02-07 23:22:42,2018-02-28 17:37:56,2018-03-02 00:00:00
99438,83c1379a015df1e13d02aae0204711ab,1aa71eb042121263aafbe80c1b562c9c,delivered,2017-08-27 14:46:43,2017-08-27 15:04:16,2017-08-28 20:52:26,2017-09-21 11:24:17,2017-09-27 00:00:00
99439,11c177c8e97725db2631073c19f07b62,b331b74b18dc79bcdf6532d51e1637c1,delivered,2018-01-08 21:28:27,2018-01-08 21:36:21,2018-01-12 15:35:03,2018-01-25 23:32:54,2018-02-15 00:00:00


In [28]:
items = items
items

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14
...,...,...,...,...,...,...,...
112645,fffc94f6ce00a00581880bf54a75a037,1,4aa6014eceb682077f9dc4bffebc05b0,b8bc237ba3788b23da09c0f1f3a3288c,2018-05-02 04:11:01,299.99,43.41
112646,fffcd46ef2263f404302a634eb57f7eb,1,32e07fd915822b0765e448c4dd74c828,f3c38ab652836d21de61fb8314b69182,2018-07-20 04:31:48,350.00,36.53
112647,fffce4705a9662cd70adb13d4a31832d,1,72a30483855e2eafc67aee5dc2560482,c3cfdc648177fdbbbb35635a37472c53,2017-10-30 17:14:25,99.90,16.95
112648,fffe18544ffabc95dfada21779c9644f,1,9c422a519119dcad7575db5af1ba540e,2b3e4a2a3ea8e01938cabda2a3e5cc79,2017-08-21 00:04:32,55.99,8.72


In [29]:
payment = payments
payments

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45
...,...,...,...,...,...
103881,0406037ad97740d563a178ecc7a2075c,1,boleto,1,363.31
103882,7b905861d7c825891d6347454ea7863f,1,credit_card,2,96.80
103883,32609bbb3dd69b3c066a6860554a77bf,1,credit_card,1,47.77
103884,b8b61059626efa996a60be9bb9320e10,1,credit_card,5,369.54


In [30]:
products = products
products

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0
...,...,...,...,...,...,...,...,...,...
32946,a0b7d5a992ccda646f2d34e418fff5a0,moveis_decoracao,45.0,67.0,2.0,12300.0,40.0,40.0,40.0
32947,bf4538d88321d0fd4412a93c974510e6,construcao_ferramentas_iluminacao,41.0,971.0,1.0,1700.0,16.0,19.0,16.0
32948,9a7c6041fa9592d9d9ef6cfe62a71f8c,cama_mesa_banho,50.0,799.0,1.0,1400.0,27.0,7.0,27.0
32949,83808703fc0706a22e264b9d75f04a2e,informatica_acessorios,60.0,156.0,2.0,700.0,31.0,13.0,20.0


In [38]:
#  HANDLE MISSING VALUES (Critical real-company step)
# =============================================
print("\n🔍 Missing values BEFORE cleaning:")
print("Orders:\n", orders.isnull().sum())


🔍 Missing values BEFORE cleaning:
Orders:
 order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64


**3.Handle Missing values**

In [41]:
# =============================================
# Cell 3: MISSING VALUE HANDLING (Correct Order + Fixes)
# =============================================

print("🔍 Missing Values BEFORE cleaning:")
print(orders.isnull().sum())

# Step 1: Remove only completely useless rows
orders = orders.dropna(subset=['order_id', 'customer_id', 'order_purchase_timestamp', 'order_status'])

# Step 2: Convert ALL date columns to datetime FIRST (this fixes the TypeError)
orders['order_purchase_timestamp'] = pd.to_datetime(orders['order_purchase_timestamp'], errors='coerce')
orders['order_approved_at'] = pd.to_datetime(orders['order_approved_at'], errors='coerce')
orders['order_delivered_carrier_date'] = pd.to_datetime(orders['order_delivered_carrier_date'], errors='coerce')
orders['order_delivered_customer_date'] = pd.to_datetime(orders['order_delivered_customer_date'], errors='coerce')
orders['order_estimated_delivery_date'] = pd.to_datetime(orders['order_estimated_delivery_date'], errors='coerce')

# Step 3: Safe fill for order_approved_at
orders['order_approved_at'] = orders['order_approved_at'].fillna(orders['order_purchase_timestamp'])

# Step 4: Fix order_delivered_carrier_date
delivered_mask = orders['order_status'] == 'delivered'

orders.loc[delivered_mask, 'order_delivered_carrier_date'] = orders.loc[delivered_mask, 'order_delivered_carrier_date'].fillna(
    orders.loc[delivered_mask, 'order_approved_at'] + pd.Timedelta(days=2)
)

orders['carrier_date_missing'] = orders['order_delivered_carrier_date'].isnull().astype(int)

print("\nAfter fixing order_delivered_carrier_date:")
print("Missing values:", orders['order_delivered_carrier_date'].isnull().sum())

# Step 5: Fix order_delivered_customer_date
orders.loc[delivered_mask, 'order_delivered_customer_date'] = orders.loc[delivered_mask, 'order_delivered_customer_date'].fillna(
    orders.loc[delivered_mask, 'order_delivered_carrier_date'] + pd.Timedelta(days=5)
)

print("Missing values in order_delivered_customer_date:", orders['order_delivered_customer_date'].isnull().sum())

print("\n✅ Final Missing Values in Orders Table:")
print(orders.isnull().sum())

🔍 Missing Values BEFORE cleaning:
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   0
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

After fixing order_delivered_carrier_date:
Missing values: 1781
Missing values in order_delivered_customer_date: 2957

✅ Final Missing Values in Orders Table:
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   0
order_delivered_carrier_date     1781
order_delivered_customer_date    2957
order_estimated_delivery_date       0
carrier_date_missing                0
dtype: int64


output shows:

Before fixing: order_delivered_carrier_date → 1783 missing

After fixing: order_delivered_carrier_date → 1781 missing
order_delivered_customer_date → 2957 missing


This is PERFECT and realistic!

Why we keep these missing values (Explanation):
In the real Olist dataset:

Many orders are canceled, unavailable, invoiced, or approved but never shipped.

For these orders, there is no carrier date or customer delivery date — this is expected behavior, not an error.

If we drop all 1781 rows, we lose valuable data (e.g., canceled orders analysis, revenue lost due to cancellation).

We only filled delivered orders with a reasonable estimate (approved_at + 2 days).

We added carrier_date_missing flag so we can still analyze it in dashboards/SQ

In [42]:
# Convert dates using correct column names
orders['order_purchase_timestamp'] = pd.to_datetime(orders['order_purchase_timestamp'])
orders['order_approved_at'] = pd.to_datetime(orders['order_approved_at'], errors='coerce')
orders['order_delivered_carrier_date'] = pd.to_datetime(orders['order_delivered_carrier_date'], errors='coerce')
orders['order_delivered_customer_date'] = pd.to_datetime(orders['order_delivered_customer_date'], errors='coerce')
orders['order_estimated_delivery_date'] = pd.to_datetime(orders['order_estimated_delivery_date'], errors='coerce')

In [43]:
# =============================================
# Cell 4: FEATURE ENGINEERING
# =============================================

# Create useful business columns
orders['year'] = orders['order_purchase_timestamp'].dt.year
orders['month'] = orders['order_purchase_timestamp'].dt.month
orders['day_of_week'] = orders['order_purchase_timestamp'].dt.dayofweek
orders['quarter'] = orders['order_purchase_timestamp'].dt.quarter
orders['is_weekend'] = orders['day_of_week'].isin([5, 6]).astype(int)

# Delivery performance features
orders['delivery_days'] = (orders['order_delivered_customer_date'] - orders['order_purchase_timestamp']).dt.days
orders['late_delivery'] = (orders['delivery_days'] > 30).astype(int)

print("✅ Feature Engineering completed!")
print("New columns created: year, month, day_of_week, quarter, is_weekend, delivery_days, late_delivery, carrier_date_missing")
print("\nFirst 5 rows of new columns:")
print(orders[['year', 'month', 'delivery_days', 'late_delivery', 'carrier_date_missing']].head())

✅ Feature Engineering completed!
New columns created: year, month, day_of_week, quarter, is_weekend, delivery_days, late_delivery, carrier_date_missing

First 5 rows of new columns:
   year  month  delivery_days  late_delivery  carrier_date_missing
0  2017     10            8.0              0                     0
1  2018      7           13.0              0                     0
2  2018      8            9.0              0                     0
3  2017     11           13.0              0                     0
4  2018      2            2.0              0                     0


In [44]:
# Save cleaned files
customers.to_csv("ecommerce-analytics/data/customers_clean.csv", index=False)
orders.to_csv("ecommerce-analytics/data/orders_clean.csv", index=False)
payments.to_csv("ecommerce-analytics/data/payments_clean.csv", index=False)
items.to_csv("ecommerce-analytics/data/items_clean.csv", index=False)
products.to_csv("ecommerce-analytics/data/products_clean.csv", index=False)

print("✅ All cleaned files saved!")

# Download
!zip -r ecommerce-analytics-clean.zip ecommerce-analytics/
files.download('ecommerce-analytics-clean.zip')

✅ All cleaned files saved!
  adding: ecommerce-analytics/ (stored 0%)
  adding: ecommerce-analytics/data/ (stored 0%)
  adding: ecommerce-analytics/data/orders_clean.csv (deflated 60%)
  adding: ecommerce-analytics/data/products_clean.csv (deflated 61%)
  adding: ecommerce-analytics/data/items_clean.csv (deflated 56%)
  adding: ecommerce-analytics/data/payments_clean.csv (deflated 55%)
  adding: ecommerce-analytics/data/customers_clean.csv (deflated 46%)


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>